ad7779 driver

ranechita opened the pull request:
Add driver for ad777x (ad7770,ad7771,ad7779) ADCs.

ranechita commented:
v2

- fixed dt-bindings error

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+		gpiod_set_value(st->reset_gpio, 1);
+		usleep_range(225, 230);
+		gpiod_set_value(st->reset_gpio, 0);
+		usleep_range(225, 230);
```
You should not need all this gpiod_set_value() calls... You should request you gpio in the asserted state (which you are doing). In the devicetree you need to specify what's the asserted state for the gpio (assuming active low in here) and so, when requesting it, you'll have your device effectively in reset. Hence, in here, you just need to bring it out of reset

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+{
+	struct ad777x_state *st = iio_priv(indio_dev);
+	int ret;
+	u8 reset_buf[8] = {0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF};
```
this is not dma safe...

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+			}
+		}
+		iio_push_to_buffers(indio_dev, &tmp[0]);
+	}
```
please look at the triggered buffer support... This really looks like it should be using it and it would make your handler simpler (eg: no need to call iio_buffer_enabled())

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+	int ret;
+	struct ad777x_state *st = ad777x_get_data(indio_dev);
+
+	disable_irq_nosync(st->spi->irq);
```
again, might seems the call to disable_irq_nosync() is not needed...

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+				    FIELD_PREP(AD777X_MOD_SPI_EN_MSK, 1));
+	if (ret)
+		return ret;
+	enable_irq(st->spi->irq);
```
same

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+			st->active_ch |= BIT(i);
+		else
+			st->active_ch &= ~BIT(i);
+	}
```
This is probably not needed... Why not using in your IRQ?

https://elixir.bootlin.com/linux/latest/source/include/linux/iio/iio.h#L583

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+			return ret;
+		return ad777x_spi_write(st,
+					AD777X_REG_CH_OFFSET_LOWER_BYTE(chan->channel),
+					lsb);
```
I would move this and IIO_CHAN_INFO_CALIBSCALE to it's own function....

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+		*val = FIELD_PREP(AD777X_UPPER, high) |
+		       FIELD_PREP(AD777X_MID, mid) |
+		       FIELD_PREP(AD777X_LOWER, low);
+		return IIO_VAL_INT;
```
ditto

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+	int gain;
+
+	if (iio_buffer_enabled(indio_dev))
+		return -EBUSY;
```
you should use this:

https://elixir.bootlin.com/linux/latest/source/drivers/iio/industrialio-core.c#L2031

And more importantly, shouldn't you have the same in ad777x_write_raw()?

ranechita replied:
When the adc enters spi sigma-delta data mode, due to the SDO line being dedicated to transmitting the data, registers can not be read, but may be written to and I thought I could leave writing permitted in the driver as well

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+		if (ret < 0)
+			return ret;
+		st->filter_enabled = mode;
+	}
```
Is this only supported in ad7771? If so, I would not expose this attr at all for the other devices...

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+{
+	struct ad777x_state *st = ad777x_get_data(indio_dev);
+
+	return st->filter_enabled;
```
Can't we read it from the device? More reliable than the cached value...

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+{
+	struct ad777x_state *st = ad777x_get_data(indio_dev);
+
+	return st->data_lines;
```
ditto

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+		return -EINVAL;
+	}
+
+	st->data_lines = mode;
```
The whole function is not safe... You need a mutex. Or likely, you also need to call in here:

https://elixir.bootlin.com/linux/latest/source/drivers/iio/industrialio-core.c#L2031

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+
+	if (readval)
+		return ad777x_spi_read(st, reg, (u8 *)readval);
+	else
```
redundant else

nunojsa commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+		ret = ad777x_spi_write(st, reg, regval);
+		if (ret)
+			return ret;
+	}
```
I see the use of crc is probably a no go for using regmap... But I would consider to add your own bus so you don't really need to doing your own ad777x_spi_write_mask() which regmap already provides (on top of other things). You can see an example of it in here:

https://elixir.bootlin.com/linux/latest/source/drivers/iio/dac/ltc2688.c#L934

ranechita replied:
I will look more into this

nunojsa reviewed (changes requested):
Ok, I'm done for round 1 of reviewing and I guess you already have plenty to do for v2 :)

amiclaus commented on Documentation/devicetree/bindings/iio/adc/adi,ad7779.yaml:
```diff
@@ -0,0 +1,89 @@
+  - clock-names
+  - vref-supply
+
+additionalProperties: false
```
`unevaluatedProperties`

amiclaus commented on Documentation/devicetree/bindings/iio/adc/adi,ad7779.yaml:
```diff
@@ -0,0 +1,89 @@
+  - |
+    #include <dt-bindings/gpio/gpio.h>
+    spi {
+        status = "okay";
```
you might want to drop the status and specify address/size cells.

amiclaus commented on Documentation/devicetree/bindings/iio/adc/adi,ad7779.yaml:
```diff
@@ -0,0 +1,89 @@
+  outputs into the full-scale ADC input range, maximizing the
+  dynamic range of the signal chain.
+
+  https://wiki.analog.com/resources/eval/user-guides/ad777x-ardz
```
i'd drop this link

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+#include "cf_axi_adc.h"
+
+#define AD777X_SPI_READ_CMD			0x80
+#define AD777X_SPI_WRITE_CMD			0x00
```
this macro is redundant.

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+
+#include "cf_axi_adc.h"
+
+#define AD777X_SPI_READ_CMD			0x80
```
BIT(7)

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+#define AD777X_SPI_READ_CMD			0x80
+#define AD777X_SPI_WRITE_CMD			0x00
+
+#define AD777X_DISABLE_SD			0x80
```
BIT(7)

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+	AD777x_4LINES,
+	AD777x_2LINES,
+	AD777x_1LINE,
+
```
nit: extra line.

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+#define AD777X_MID				GENMASK(15, 8)
+#define AD777X_LOWER				GENMASK(7, 0)
+
+#define AD777X_CRC8_POLY	0x07
```
indent properly please.

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+
+	st = iio_priv(indio_dev);
+
+	st->vref = devm_regulator_get(&spi->dev, "vref");
```
ad7779 has an internal reference of 2.5V. therefore you might want to use `devm_regulator_get_optional`

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+		return IIO_VAL_INT;
+	}
+
+	return -EINVAL;
```
drop this and use the default case in the switch. check for similar approaches in the entire code.

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+				    AD777X_MOD_POWERMODE_MSK,
+				    FIELD_PREP(AD777X_MOD_POWERMODE_MSK, mode));
+
+	if (ret < 0)
```
`(ret)` ?

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+	 * DMA (thus cache coherency maintenance) requires the
+	 * transfer buffers to live in their own cache lines.
+	 */
+	u8			reg_rx_buf[3] ____cacheline_aligned;
```
not sure if we have the `IIO_DMA_MINALIGN` definition in our tree, but you might want to consider `____cacheline_aligned` -> `__aligned(IIO_DMA_MINALIGN)`

ranechita replied:
it isn't in our tree, but if it is seems fit I can do something like this:
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/addac/ad74115.c#L43

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+		return PTR_ERR(st->mclk);
+
+	ret = clk_prepare_enable(st->mclk);
+	if (ret < 0)
```
ditto `(ret)`

amiclaus commented on Documentation/devicetree/bindings/iio/adc/adi,ad7779.yaml:
```diff
@@ -0,0 +1,89 @@
+  - reg
+  - clocks
+  - clock-names
+  - vref-supply
```
this should be dropped from here, taking into account the internal reference option.

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+	u8 exp_crc = 0;
+	struct spi_transfer reg_read_tr[] = {
+		{
+			.tx_buf = st->reg_tx_buf,
```
more like a question: can't you use the same `st->reg_buf` for both `tx_buf` and `rx_buf` members of the spi_transfer function?

ranechita replied:
I'm not entirely sure how the buffers are being handled but I think it may be unsafe and lead to inconsistent data on tx. I've seen the same approach consistently in other drivers as well

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+	if (st->power_mode == AD777X_LOW_POWER)
+		div = AD777X_LOWPOWER_DIV;
+	else
+		div = AD777X_HIGHPOWER_DIV;
```
nit: new line after this

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+					AD777X_REG_CH_GAIN_LOWER_BYTE(chan->channel),
+					lsb);
+	case IIO_CHAN_INFO_CALIBBIAS:
+		msb = (val >> 16) & 0xFF;
```
`FIELD_GET`

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+					lsb);
+	case IIO_CHAN_INFO_CALIBBIAS:
+		msb = (val >> 16) & 0xFF;
+		mid = (val >> 8) & 0xFF;
```
ditto.

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+	case IIO_CHAN_INFO_CALIBBIAS:
+		msb = (val >> 16) & 0xFF;
+		mid = (val >> 8) & 0xFF;
+		lsb = val & 0xFF;
```
ditto

amiclaus reviewed (commented):
some initial comments on my side. 2 minor general aspects:
1. the drivers supports ad7770, ad7771 and ad7779. so why naming all the stuff ad7779 instead of ad7770 which is the first part in the list?
2. there is a checkpatch warning that can be addressed.

ranechita commented:
v3:

- removed power_mode from ext info and added it in pm_ops
- added of_table and chip_info
- addressed styling issues
- removed filter_type attr from devices that do not support it
- switched to get_optional where it was the case
- added Documentation/ABI/testing file for ext_info attrs
- removed redundant code
- switched to FIELD_GET where it was the case
- changed additionalProperties -> unevaluatedProperties
- moved calibbias & calibscale read/write to separate functions

ranechita commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1043 @@
+					      &ad777x_buffer_setup_ops,
+					      NULL);
+	if (ret)
+		return dev_err_probe(&st->spi->dev, ret,
```
Yes, the interrupt occurs at the end of each conversion. I was also unsure whether it should be a trigger_buffer or kfifo but other some discussions I thought it makes more sense to implement it like this, using the irq handler, given that the interrupt indicates new data is available and I've noticed a similar approach upstream as well. Ultimately, it wouldn't be a big change to switch to triggered buffer if it is more fit.

ranechita commented:
v4:

- addressed one checkatch warning

Regarding the naming, I initially started the development on ad7779 and only later I found out about the other two parts in the family, so I didn't change the name of the files. Should it be switched to ad7770/ad777x?

amiclaus commented on drivers/iio/adc/ad7779.c:
```diff
@@ -0,0 +1,1183 @@
+		return ret;
+
+	if (spi->irq)
+		ret = ad777x_register(st, indio_dev);
```
Not sure about the current state of the upstream axi-adc driver (if it is up to date or not), but did you consider sending the non-axi version upstream? Might simplify the review process.
Afterwards you can add a separate patch in our tree for the axi-adc support that should be pretty straightforward.

nunojsa commented:
> Not sure about the current state of the upstream axi-adc driver (if it is up to date or not), but did you consider sending the non-axi version upstream? Might simplify the review process.
Afterwards you can add a separate patch in our tree for the axi-adc support that should be pretty straightforward.

Please don't use the upstream axi-adc... It's also fairly broken and I intend to start work on it in the next few weeks (if nothing else comes). But If there's a "non axi" version that's upstreamable, please go for it.

ranechita commented:
If everything is considered ok on this version I will split the commits and send the "non-axi" version upstream, but of there are other issues to be addressed I would do that first to make sure all is in order first

Siegurd01 commented:
Is there a chance to integrate AD7779 driver for BeagleBone Black?

nunojsa commented:
@ranechita , what's the status for this PULL? Did you ever sent the patches upstream?

ranechita commented:
> @ranechita , what's the status for this PULL? Did you ever sent the patches upstream?

No, but I will rebase it and split the commits, and then send it upstream. Until then if there are further reviews on the code, I can fix those first

ranechita commented:
v6:

- rebased with main and split the commits in order to send non-axi version upstream

nunojsa commented:
@ranechita this seems to be already upstream? Can you update this PR?

ranechita commented:
@nunojsa The upstream version is without axi, I will upstream a new patch that uses iio-backend for axi and it will added to a different branch. Is it ok if I close this pull request in this case?

nunojsa commented:
> @nunojsa The upstream version is without axi, I will upstream a new patch that uses iio-backend for axi and it will added to a different branch. Is it ok if I close this pull request in this case?

Do you have any updates on this? If you think it makes sense, please go ahead and close this PR
